### ISSUE TITLE
\finrl\config.py which parameter is the best for apply to algorithm? How can i choose?

### ISSUE THREAD

**Issue by omerfirat** [AI4Finance-Foundation/FinRL#1062]
I have defined specific ranges for each hyperparameter, and I want to find the best parameter for A2C or other algorithms. However, there can be numerous combinations, and how can I find the best parameter? Is it through trial and error? Even if I find the best parameter once, when I retrain with the same parameter, it produces different results. This is inherent in the nature of models. How can I fix the right parameter? It requires a lot of trial and error to achieve this.

I have increased and decreased the values of these added hyperparameters by 1. How much should I increase or decrease them? Even though my machine is powerful, it takes a lot of time for huge parameter like 10 params for each hyper prams. Can you help me reach the best parameter values?"

    n_steps_values       = [4, 5, 6]
    ent_coef_values      = [0.001, 0.01, 0.1]
    learning_rate_values = [0.00007, 0.0007, 0.007]

I need this information because I am working on a thesis for this topic

**Comment by mmmarchetti** [comment]
Certainly, hyperparameter tuning for reinforcement learning (RL) algorithms can be challenging due to the inherent variability in results. Here are some steps and recommendations you can follow to address this:

1. **Random Search Over Grid Search**: 
   - Traditional grid search, where you try every combination of hyperparameters, can be expensive, especially when the number of hyperparameters is large.
   - Research has shown that random search can be more efficient than grid search. Instead of exhaustively trying every combination, you sample hyperparameter combinations at random and test them. Even with fewer trials, you can often find a good combination.

2. **Bayesian Optimization**:
   - Bayesian optimization is a probabilistic model-based optimization technique. It's especially useful for optimizing expensive-to-evaluate functions like the ones in RL.
   - Libraries like [Optuna](https://optuna.org/), [Hyperopt](http://hyperopt.github.io/hyperopt/), and [BayesianOptimization](https://github.com/fmfn/BayesianOptimization) can be used to implement Bayesian optimization for hyperparameter tuning.
   
3. **Evolutionary Algorithms**:
   - Another approach is to use evolutionary algorithms to evolve the best set of hyperparameters. There are frameworks like [Nevergrad](https://github.com/facebookresearch/nevergrad) that can assist with this.

4. **Start With Fewer Episodes**:
   - Instead of running full training during each hyperparameter trial, you can run a fraction of the episodes to get a rough estimate of performance. Once you've narrowed down a few promising hyperparameters, you can perform more extensive training.

5. **Use a Seed**:
   - To address the variability in results, you can set a random seed. This way, the random processes (like network initialization and action selection) will be the same between runs. However, remember that a model that is only good for a particular seed may not be generally good. It's a trade-off between reproducibility and generality.

6. **Learning Rate Schedulers**:
   - Instead of using a fixed learning rate, consider using learning rate schedulers which adjust the learning rate during training. This might help in better convergence.

7. **Parallelize**:
   - If possible, run multiple training sessions in parallel, especially if you have access to multiple CPUs or GPUs. This will allow you to test multiple hyperparameters simultaneously.

8. **Incremental Changes**:
   - It's not always necessary to test a wide range of values for each hyperparameter immediately. Start with a reasonable guess based on literature or intuition, and make small, incremental changes to see its effect.

Given the provided hyperparameters:

- `n_steps_values` = [4, 5, 6]
- `ent_coef_values` = [0.001, 0.01, 0.1]
- `learning_rate_values` = [0.00007, 0.0007, 0.007]

You have \(3 \times 3 \times 3 = 27\) combinations if you were to do a full grid search. If you do random sampling, you might sample, for instance, 10 of these combinations at random, which would reduce the computational burden.

Lastly, remember that the best hyperparameters often depend on the specific environment and problem you're working with. It's not uncommon to have to re-tune hyperparameters when you switch to a different task or environment.
